bee-network/src/swarm/protocols/iota_gossip/handler.rs

```diff
@@ -0,0 +1,205 @@
+
+        let IotaGossipHandlerInEvent { origin } = incoming_event;
+
+        // We only send the upgrade request if this handler belongs to an outbound connection.
+        if origin == Origin::Outbound {
+            let send_request = ProtocolsHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(IotaGossipProtocolUpgrade::new(self.info.clone()), ()),
+            };
+
+            debug!("gossip handler: sending protocol upgrade request.");
+
+            self.events.push_back(send_request);
+        }
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Injects the output of a successful upgrade on a new inbound substream.
+    fn inject_fully_negotiated_inbound(&mut self, new_inbound: NegotiatedSubstream, _: Self::InboundOpenInfo) {
+        let negotiated_inbound = ProtocolsHandlerEvent::Custom(IotaGossipHandlerEvent::UpgradeCompleted {
+            substream: Box::new(new_inbound),
+        });
+
+        debug!("gossip handler: fully negotiated inbound.");
+
+        self.events.push_back(negotiated_inbound);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Injects the output of a successful upgrade on a new outbound substream.
+    ///
+    /// The second argument is the information that was previously passed to
+    /// [`ProtocolsHandlerEvent::OutboundSubstreamRequest`].
+    fn inject_fully_negotiated_outbound(&mut self, new_outbound: NegotiatedSubstream, _: Self::OutboundOpenInfo) {
+        let negotiated_outbound = ProtocolsHandlerEvent::Custom(IotaGossipHandlerEvent::UpgradeCompleted {
+            substream: Box::new(new_outbound),
+        });
+
+        debug!("gossip handler: fully negotiated outbound.");
+
+        self.events.push_back(negotiated_outbound);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Notifies the handler of a change in the address of the remote.
+    fn inject_address_change(&mut self, new_address: &Multiaddr) {
+        debug!("gossip handler: new address: {}", new_address);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Indicates to the handler that upgrading an outbound substream to the given protocol has failed.
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _: Self::OutboundOpenInfo,
+        e: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Error>,
+    ) {
+        debug!("gossip handler: outbound upgrade error: {:?}", e);
+
+        // self.events.push_back(ProtocolsHandlerEvent::Close(e));
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Indicates to the handler that upgrading an inbound substream to the given protocol has failed.
+    fn inject_listen_upgrade_error(
+        &mut self,
+        _: Self::InboundOpenInfo,
+        e: ProtocolsHandlerUpgrErr<<Self::InboundProtocol as InboundUpgradeSend>::Error>,
+    ) {
+        debug!("gossip handler: inbound upgrade error: {:?}", e);
+        // let err = match e {
```

Comment:
Why is this commented?

Reply:
I want to finish this impl in the next patch version. I don't have time for this atm.